qcodes/instrument_drivers/american_magnetics/AMI430.py

```diff
@@ -0,0 +1,618 @@
+                           vals=Numbers())
+
+        self.add_parameter('field',
+                           get_cmd=self._get_field,
+                           set_cmd=self._set_field,
+                           units='T',
+                           vals=Numbers())
+
+    def _get_angle_offset(self):
+        return np.degrees(self._angle_offset)
+
+    def _set_angle_offset(self, angle):
+        # Adjust the field if the offset angle is changed
+        if self._angle_offset != np.radians(angle):
+            self._angle_offset = np.radians(angle)
+            self._set_field(self._field)
+
+    def _get_angle(self):
+        angle = np.arctan2(self.magnet_y.field(), self.magnet_x.field())
+
+        return np.degrees(angle - self._angle_offset)
+
+    def _set_angle(self, angle):
+        self._angle = np.radians(angle)
+
+        self._set_field(self._field)
+
+    def _get_field(self):
+        x = self.magnet_x.field()
+        y = self.magnet_y.field()
+
+        return np.sqrt(x**2 + y**2)
+
+    def _set_field(self, field):
+        self._field = field
+
+        B_x = field * np.cos(self._angle + self._angle_offset)
+        B_y = field * np.sin(self._angle + self._angle_offset)
+
+        # First ramp the magnet that is decreasing in field strength
+        if np.abs(self.magnet_x.field()) < np.abs(B_x):
+            self.magnet_x.field(B_x)
+            self.magnet_y.field(B_y)
+        else:
+            self.magnet_y.field(B_y)
+            self.magnet_x.field(B_x)
+
+
+class AMI430_3D(Instrument):
+    """
+    Virtual driver for a system of three AMI430 magnet power supplies.
+    This driver provides methods that simplify setting fields in
+    different coordinate systems.
+
+    Cartesian, spherical and cylindrical coordinates are supported, with
+    the following parameters:
+
+    Carthesian:     x,      y,      z
+    Spherical:      phi,    theta,  field
+    Cylindrical:    phi,    rho,    z
+
+    For the spherical system theta is the polar angle from the positive
+    z-axis to the negative z-axis (0 to pi). Phi is the azimuthal angle
+    starting at the positive x-axis in the direction of the positive
+    y-axis (0 to 2*pi).
+
+    In the cylindrical system phi is identical to that in the spherical
+    system, and z is identical to that in the cartesian system.
+
+    All angles are set and returned in units of degrees and are automatically
+    phase-wrapped.
+
+    If you want to control the magnets in this virtual driver individually,
+    one can set the _parent_instrument parameter of the magnet to None.
+    This is done at your own risk, as it skips field strength checks and
+    might result in a magnet quench.
+
+    Example of instantiation:
+
+    magnet = AMI430_3D('AMI430_3D',
+        AMI430('AMI430_X', '192.168.2.3', 0.0146, 68.53, 0.2),
+        AMI430('AMI430_Y', '192.168.2.2', 0.0426, 70.45, 0.05),
+        AMI430('AMI430_Z', '192.168.2.1', 0.1107, 81.33, 0.08))
```

Comment:
the field limit is not in this example, which is a required positional argument for AMI430_3D